Remove I/O pool blocking sniffing call from onFailure callback, add some logic around host exclusion

reembs opened the pull request:
Addressing: https://github.com/elastic/elasticsearch/issues/27984

In addition to the bugfix, I tried making the mechanism a bit clearer in a mutlithread context. Beforehand the onFailure execution stack would call the sniffing code in the callback (bad). In that context it kinda made sense to exclude a single host, but since we're now delaying the execution (and because other requests might be running concurrently) it's clearer to use a list of hosts that need to be excluded.

This list expires its values after a certain number of sniffing cycles (sniffing calls) had passed. This number defaults to 1 to mimic the existing behavior more closely.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

DaveCTurner commented:
@elasticmachine please test this

reembs commented:
addressing earlier issue from July reported by @ssesha https://github.com/elastic/elasticsearch/issues/25701

@DaveCTurner I fixed the styling issues and added documentation to the new public method. Can you trigger the build again?

DaveCTurner commented:
@elasticmachine please test this

DaveCTurner commented:
@reembs I can't recall what the styling issue was now, and it looks like you rewrote history to remove the commit that I looked at earlier so any comments I made are now lost. It makes it easier on reviewers if you just add commits to your branch. We squash the whole branch when it's merged so don't worry about making a noisy log at this stage. Please could you avoid rewriting history in future?

I have no further comments on the code, so over to @javanna for a proper review as he knows this area better than I do.

reembs commented:
@DaveCTurner sorry about that habit. Issue was line length, I think. Your comments were on the PR so they are all still here.

s1monw commented:
@javanna or @tlrx can you take a look at this please? @reembs sorry for the delay

tlrx commented:
I'd be happy to have @javanna 's opinion too

reembs commented:
> Instead, I think that the Sniffer.Task.sniffOnFailure() method could use its existing ScheduledExecutorService to schedule a new round of sniffing but this time removing the failed host (at it is already the case, but in a non blocking way using the ScheduledExecutorService).

@tlrx I get what you're saying about moving the the async `ScheduledExecutorService` call from `Sniffer.sniffOnFailure()` to `Sniffer.Task.sniffOnFailure()`, from an encapsulation perspective. Regarding the single/many host exclusion issue, I'm not sure I understand the reasoning of excluding just a single host. If the REST Client already has a dead-host concept, no host exclusion should be done on the Sniffer level. If we do choose to exclude hosts independently in the Sniffer, especially since the sniffing call on-failure is now async, excluding just a single host seems arbitrary. The PR changes the behavior from "exclude last host that failed" to "exclude all hosts that failed since the last sniffing cycle", which is a more reasonable approach IMO considering that cluster sizes often change in increments larger than 1.

tlrx commented:
> @tlrx I get what you're saying about moving the the async ScheduledExecutorService call from Sniffer.sniffOnFailure() to Sniffer.Task.sniffOnFailure(), from an encapsulation perspective.

Good! I think this is the main point of this pull request, ie fix the bug that blocks the execution. That would be great if we can come up with a test for this too.

> Regarding the single/many host exclusion issue, I'm not sure I understand the reasoning of excluding just a single host.

Sorry if I wasn't clear enough. I didn't mean to have a reasoning about excluding just a single host. This is true that the Sniffer adds back all the nodes it just sniffed except the one that just failed. But this is also true that the `RestClient` already has the logic of taking care of failing nodes and adding them back, so I think we must not duplicate a similar logic at the Sniffer level but instead we could (thinking out loud) expose a DeadHostStrategy/ReconnectPolicy/Whatever that would indicate when a failed node can be retried -  whether because a given time as elapsed or because it failed during N sniffing rounds etc

tlrx commented:
@reembs Would you like to revisit this?

reembs commented:
I would like to get this merged, yes. Can we decide on host exclusion logic in context of this PR and bugfix?

Currently the PR introduces miltiple host exclusions on the sniffer level.

Since the PR is a bugfix and isn't meant to modify behavior, a reasonable solution would be to revert back to single host exclusion in the sniffer, although I find that strange in a mutli threaded object like the sniffer.

Second option is the one in the PR currently, multiple exclusions for all failing hosts from the previous cycle.

A thind option would be to remove host exclusion from the sniffer altogether.

Lets decide on one of these and I'll be happy to implement.

IMO, more sophisticated methods are outsude the scope of this bugfix.

@tlrx

javanna commented:
heya @reembs, first of all sorry, once again, because it took me ages to get to this. I see the bug and your fix is good, I would love to get it in. Thanks a lot for taking the time to debug the problem and send a fix.

I just read the comments above. It is true that the REST client has its own blacklist like @tlrx said above. When sniffing is enabled though, every time the sniffer runs it replaces the hosts with the sniffed ones and clears the blacklist. The idea is that when nodes are listed in the cluster state, they are available. The blacklist though is maintained based on availability of nodes since the next sniffing round, assuming that nodes that are unresponsive will also fall out of the cluster. 

If you sniff on failure, it may happen that you get a failure for a node that is still part of the cluster hence returned when sniffing. You then want to filter that one node out as you know that it just failed, otherwise it will be immediately retried which causes more failures which causes more sniffing rounds etc. Let's say that when using sniff on failure, the blacklist becomes less important, as we will re-sniff for each single failure which means replacing the hosts and clearing the blacklist every time. 

Why one single host? The assumption is that if that host fails, we want to make sure not set it back to the client even though still part of the cluster state, and it can only be set back later if the following sniff run (which will happen with a different delay, called `sniffAfterFailureDelay`) will again return that node, which should happen only once the node is back. If another node fails, the same happens but the previous failed nodes should not be returned as part of the cluster state unless back alive, assuming that enough time has passed and the node has fallen out of the cluster. To summarize, because sniffing on failure kind of overrides the internal blacklist, filtering out the failed node immediately addresses the case where the cluster didn't realize yet that such node is failing. But nothing more should be needed if you sniff at every failure. 

I would appreciate if you could update your PR and remove the new list of hosts to be excluded. Also, although the sniffer is not well tested today (I am working on improving this), it would be great to have a test for this. Do you think that you could reproduce the bug either in a unit test or sending requests to a mock http server like we do in other rest client tests?

javanna commented on client/sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java:
```diff
@@ -106,35 +119,63 @@ synchronized void scheduleNextRun(long delayMillis) {
                     }
                 } catch (Exception e) {
                     logger.error("error while sniffing nodes", e);
+                    nextSniffDelay = sniffAfterFailureDelayMillis;
```
The idea behind `sniffAfterFailureDelay` is that once a node fails, you want to re-sniff, but you may also want to check if the node has come back through a new sniffing round earlier than you'd usually do. So in general `sniffAfterFailureDelay` would be lower than `sniffInterval`. I see that here you are using now a different interval for cases where sniffing itself has failed, but that was not the purpose of the `sniffAfterFailureDelay`. It is not a bad idea to actually re-sniff earlier if something goes wrong, but it should be a different concept.

javanna commented on client/sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java:
```diff
@@ -53,15 +56,16 @@
     public void sniffOnFailure(HttpHost failedHost) {
-        this.task.sniffOnFailure(failedHost);
+        this.task.failedHosts.putIfAbsent(failedHost, 0L);
+        this.task.scheduleNextRun(0);
```
I suspect that having to pass the next interval which differs from the usual interval is the reason why this was not async in the first place :) we need to find a way to do that though, we currently lose that behaviour with this change.

Another idea, maybe overkill, could be to change the hosts sniffer API to be async and accept a listener as an argument.

javanna commented:
@reembs thanks again for your PR and diving into this one!

Based on my comments above, and some recent work I did in this area, I would close this as it's now superseded by #29638 which fixes the original issue and adds tests for sniffing.

javanna commented:
one more thing, I have incorporated some of your concerns around taking out one host when sniffing on failure. We discussed this internally and we went for taking no hosts out after all and rely entirely on the returned nodes from the cluster. Taking one host out was proving tricky to test and debug, and probably not that useful either.
